Support streaming in dev-ui chat

iocanel opened the pull request:
The pull request introduces a new option (streaming chat). The option is enabled when model supports streaming chat model and rag is disabled (streaming chat can't work with rag atm). When user enables streaming chat. Responses from the model are streamed to the devui chat screen.

- Resolves: #442

jmartisk commented:
I'm at a conference so I'll have time to test it later, but I'll just throw in that, in case of a timeout, or whatever other error from the LLM, are we sure that we don't get the history in the Dev UI desynchronized with the one stored in the actual chat memory? That's why the original synchronous variant of the `chat` method always returns the whole history - so the Dev UI can replace the whole history by that, and prevent desynchronization. I think this might be a little tricky to implement robustly when you just stream a single message.

iocanel commented:
> I'm at a conference so I'll have time to test it later, but I'll just throw in that, in case of a timeout, or whatever other error from the LLM, are we sure that we don't get the history in the Dev UI desynchronized with the one stored in the actual chat memory? That's why the original synchronous variant of the `chat` method always returns the whole history - so the Dev UI can replace the whole history by that, and prevent desynchronization. I think this might be a little tricky to implement robustly when you just stream a single message.

The memory gets modified when `onComplete` is called, with the whole message. In case of an error the memory is restored from the backup in a way similar to regular chat. Do you think we need to do more?

geoand commented:
I plan on going to do a `0.14.0` today, so we can leave this for `0.14.1` once @jmartisk gives his final approval

jmartisk commented on core/deployment/src/main/resources/dev-ui/qwc-chat.js:
```diff
@@ -96,6 +121,15 @@ export class QwcChat extends LitElement {
+                                      this.render();
+                                  }}"/></div>
+            <div><vaadin-checkbox label="Enable Streaming Chat (if supported by model & rag disabled)"
+                                  class="${this.jsonRpc.isStreamingChatSupported() ? 'show' : 'hide'}"
```
This jsonrpc method should be called only once - this way you call it again every time the form is rendered

iocanel commented:
I updated the PR and added error handling in all possible ways I could imagine:

- using `orError`
- handling the error as part of `onNext`
- wrapping the whole thing in try / catch

Even though the error is reaching the UI as I can tell from the console, I can't somehow capture it in my javascript code.
@phillip-kruger any idea what I am doing wrong here?

iocanel commented:
@jmartisk: I checked what you do in `qwc-chat.js` and it's not applicable to this case. In this case json rpc returns a multi errors are not just catch using `catch`. They should be caught using `onError` but the errors are not propagated. @phillip-kruger mentioned that it most probably an issue in json rpc impl.

phillip-kruger commented:
I have not looked at this yet, but in the mean time, can you see if anything prints in :
1) The browser console ?
2) The dev-ui log (the json message tab)

jmartisk commented:
I'm now realizing, this will also cause problems when Tools are used, not just RAG, no? Because the streaming response from the server will not contain tool-related messages. :(
To fix that, we would have to check the chat memory for changes (for the tool messages to appear in it) while the streaming is still going on, and somehow hack them into the stream (for sending the augmented message with RAG, we'd actually have to do something similar).

geoand commented:
Yeah, that sounds right

iocanel commented:
@jmartisk: the issue you had with errors being swallowed has been fixed on the quarkus side of the house by @phillip-kruger and should be available in 3.11.1. Nothing additional needed here.

Now, regarding your comment about tools, the suggested solution is not clear to me. Is it something that is required for this PR or something we could add later on?

jmartisk commented:
It would be nice to have RAG and tools working properly with streaming, but if you don't want to do it I can have a look. I have some heavier refactoring planned anyway, to be able to support images.
